src/PackageManagement/IDE/DeleteOnRestartManager.cs

```diff
@@ -0,0 +1,216 @@
+            foreach (var directory in filesWithoutFolders)
+            {
+                File.Delete(directory + DeletionMarkerSuffix);
+            }
+
+            return candidates.Where(path => Directory.Exists(path)).ToList();
+        }
+
+        /// <summary>
+        /// Checks for any pacakge directories that are pending to be deleted and raises the
+        /// <see cref="PackagesMarkedForDeletionFound"/> event.
+        /// </summary>
+        public virtual void CheckAndRaisePackageDirectoriesMarkedForDeletion()
+        {
+            var packages = GetPackageDirectoriesMarkedForDeletion();
+            if (packages.Any() && PackagesMarkedForDeletionFound != null)
+            {
+                var eventArgs = new PackagesMarkedForDeletionEventArgs(packages);
+                PackagesMarkedForDeletionFound(this, eventArgs);
+            }
+        }
+
+        /// <summary>
+        /// Marks package directory for future removal if it was not fully deleted during the normal uninstall process
+        /// if the directory does not contain any added or modified files.
+        /// The package directory will be marked by an adjacent *directory name*.deleteme file.
+        /// </summary>
+        [SuppressMessage(
+            "Microsoft.Design",
+            "CA1031:DoNotCatchGeneralExceptionTypes",
+            Justification = "We want to log an exception as a warning and move on")]
+        public void MarkPackageDirectoryForDeletion(
+            PackageIdentity package,
+            string packageRoot,
+            INuGetProjectContext projectContext)
+        {
+            if (PackagesFolderPath == null)
+            {
+                return;
+            }
+
+            try
+            {
+                using (FileSystemUtility.CreateFile(packageRoot + DeletionMarkerSuffix, projectContext))
+                {
+                }
+            }
+            catch (Exception e)
+            {
+                projectContext.Log(
+                    MessageLevel.Warning,
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.Warning_FailedToMarkPackageDirectoryForDeletion,
+                        packageRoot,
+                        e.Message));
+            }
+        }
+
+        /// <summary>
+        /// Attempts to remove package directories that were unable to be fully deleted during the original uninstall.
+        /// These directories will be marked by an adjacent *directory name*.deleteme files in the local package
+        /// repository.
+        /// If the directory removal is successful, the .deleteme file will also be removed.
+        /// </summary>
+        [SuppressMessage(
+            "Microsoft.Design",
+            "CA1031:DoNotCatchGeneralExceptionTypes",
+            Justification = "We want to log an exception as a warning and move on")]
+        public void DeleteMarkedPackageDirectories(INuGetProjectContext projectContext)
```

Comment:
when the solution Gets Opened/Closed & When the NugetProject package gets loaded ( i.e when the powershell command prompt is opened or PackageManager UI is opened.).